### PR TITLE
Insert field into config

### PR DESCRIPTION
**Pull request by samwestmoreland**

### CODE REVIEW

File: ast/ast.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+//
+type File struct {
+	// set  *FileSet
+	Name string // file name as provided to AddFile
+	Base int    // Pos value range for this file is [base...base+size]
+	Size int    // file size as provided to AddFile
+
+	// lines and infos are protected by set.mutex
+	Lines       int
+	NumSections int
```
Comment (Tatskaari) [inline ast/ast.go]: nit: We could just `len(File.Sections)`.

File: ast/ast.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	// Sections    map[string]Section
+	Sections []Section
+}
+
+type Section struct {
+	Title             string
+	Key               string
+	Line              int
+	Fields            []Field
+	LeadingWhiteSpace int
```
Comment (Tatskaari) [inline ast/ast.go]: This doesn't seem to be used?

File: ast/ast.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	if len(split) != 2 {
+		log.Fatalf("Got invalid field \"%v\"", s)
+	}
+	return Field{
+		Key:   split[0],
+		Value: split[1],
+	}
+}
+
+// Strip brackets from section header and lower,
```
Comment (Tatskaari) [inline ast/ast.go]: nit: Go standard is to have this in the format `// FunctionName does x` i.e. `GetSectionKeyFromString strips brackets from....`

File: ast/read.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			f.NumLines += 1
+		} else if matched, err := regexp.MatchString(`^ *\[.*\] *$`, line); err != nil {
+			log.Panicf("Error matching regexp: %v", err)
+		} else if matched {
+			f.Sections = append(f.Sections, MakeSection(line))
+			currentSection = f.Sections[len(f.Sections)-1].Key
+			f.NumLines += 1
+		} else {
+			f.NumLines += 1
+		}
```
Comment (tiagovtristao) [inline ast/read.go]: I think `f.NumLines += 1` can be moved down after `else` for simplification.

Comment (tiagovtristao) [inline ast/read.go]: How does this bit work? This is reached when there isn't a regex error, and a match wasn't found. I guess you want to error out in this case? For instance, if a line like `I'm a random line` is invalid.

File: ast/ast.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+func MakeField(s string) Field {
+	if s == "" {
+		// A field can be a blank line
+		return Field{}
+	} else if strings.HasPrefix(s, ";") || strings.HasPrefix(s, "#") {
+		// or a comment
+		return Field{
+			Key:     "",
+			Value:   "",
```
Comment (tiagovtristao) [inline ast/ast.go]: nit: No need to set. Go will zero-value them by default.

File: ast/modify.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+			f.Sections[i].Fields = append(f.Sections[i].Fields, field)
+			f.NumLines += 1
+			return f
+		}
+	}
+
+	// Couldn't find section so create new one
+	// Check if file currently ends with a blank line first
+	needToAppendSpace := true
+	lastSection := &f.Sections[len(f.Sections)-1]
```
Comment (tiagovtristao) [inline ast/modify.go]: Unlikely, but an out-of-bounds error will occur if the file is empty.

File: ast/modify.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	if len(lastSection.Fields) > 0 {
+		lastField := &lastSection.Fields[len(lastSection.Fields)-1]
+		if lastField.IsBlankLine() {
+			needToAppendSpace = false
+		}
+	}
+
+	if needToAppendSpace {
+		lastSection.Fields = append(lastSection.Fields, Field{Key: "", Value: ""})
```
Comment (tiagovtristao) [inline ast/modify.go]: nit: `Field{}`

File: ast/ast_test.go [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+	file := Read(strings.NewReader(config), "test")
+	require.Equal(t, len(file.Sections), 2)
+	require.Equal(t, file.Sections[1].Key, "rosaceae \"subsection\"")
+	require.Equal(t, file.Sections[1].Header, "[Rosaceae \"subsection\"]")
+
+	field := Field{
+		Key:   "Malus prunifolia",
+		Value: "Chinese crabapple",
+	}
+	file = InjectField(file, field, "rosaceae \"subsection\"", true)
```
Comment (tiagovtristao) [inline ast/ast_test.go]: From a public API perspective, I think it's preferable to offer something like `InjectField(file, field, "rosaceae", "subsection", true)`, then inside the function we can construct it as we wish.

**Review by tiagovtristao — COMMENTED** [review]
Thorough testing!

File: ast/ast.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+// A File is an AST representation of a config file.
+type File struct {
+	name     string
+	sections []section
+}
+
+type section struct {
+	// str is the as-read section if the
+	// section was read from a config file
+	str        string
```
Comment (Tatskaari) [inline ast/ast.go]: Ah I see! I think this is more like the section heading. When I first read this I thought it was the entire section. What do you think of calling it `headingStr`?

**Review by Tatskaari — APPROVED** [review]
Looks good to me!

**Review by Tatskaari — APPROVED** [review]

**Pull request merged**